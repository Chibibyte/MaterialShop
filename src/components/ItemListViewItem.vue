<template>
  <v-hover>
    <template v-slot="{ hover }">
      <v-card hover :elevation="hover ? 16 : 4" ripple class="pa-0 ma-0">
        <v-row :class="'ma-0' + (flexRow ? '' : ' flex-column')">
          <v-col class="pa-0">
            <v-container class="blue d-flex flex-column ma-0 pa-0">
              <v-icon right class="secondary ma-0 white--text" size="96px"
                >mdi-{{ item.title }}</v-icon
              >
            </v-container>
          </v-col>
          <v-col class="pa-0">
            <v-card-title>{{ item.title }}</v-card-title>
            <v-card-subtitle>{{ item.price }}â‚¬</v-card-subtitle>
            <v-row justify="space-around">
              <v-col :cols="hideButtons ? 10 : 4">
                <v-select
                  ref="vselect"
                  :disabled="disabled"
                  :readonly="readonly"
                  v-model="amount"
                  dense
                  :items="amountArr"
                  label="Amount"
                  @change="() => updateItem({ id: item.id, amount })"
                ></v-select>
              </v-col>
              <v-col v-if="!clearingCart && !hideButtons" cols="6">
                <v-btn @click="addItem({ id: item.id, amount })">
                  <v-icon left>mdi-cart-arrow-down</v-icon>
                  <span>Add</span>
                </v-btn>
              </v-col>
              <v-col v-else-if="!hideButtons" cols="6">
                <v-btn @click="deleteItem(item.id)">
                  <v-icon left>mdi-cart-arrow-down</v-icon>
                  <span>Remove</span>
                </v-btn>
              </v-col>
            </v-row>
          </v-col>
        </v-row>
      </v-card>
    </template>
  </v-hover>
</template>

<script>
/**
 * Business Logic
 *    -represents a single item in the ItemListView
 *    -display:
 *      -material item
 *      -material item name
 *      -price
 *      -amount
 *      -add button OR
 *      -delete button
 *
 *    -the user can change the amount in the amount selection
 *    -clicking on the add button sends a call to the addItem function in the noncheckout module of the vuex store
 *    -clicking on the delete button sends a call to the deleteItem function in the noncheckout module of the vuex store
 *
 * Inputs:
 * Outputs:
 */
export default {
  name: "ItemListViewItem",
  props: {
    updateOnChange: {
      type: Boolean,
      default: false,
    },
    item: {
      type: Object,
      default() {
        return {
          title: "DEFAULT TITLE",
          price: "-1",
          id: "0",
        };
      },
    },
    flexRow: {
      type: Boolean,
      default: false,
    },
    initAmount: {
      type: Number,
      default: 1,
    },
    clearingCart: {
      type: Boolean,
      default: false,
    },
    hideButtons: {
      type: Boolean,
      default: false,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    readonly: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      amountArr: Array.from(Array(30).keys()).map((v) => v + 1),
      amount: 0,
    };
  },
  methods: {
    addItem(item) {
      this.$store.commit("noncheckout/addItem", item);
    },
    deleteItem(itemId) {
      this.$store.commit("noncheckout/deleteItem", itemId);
    },
    updateItem(item) {
      this.$store.commit("noncheckout/updateItem", item);
    },
  },
  beforeMount() {
    this.amount = this.initAmount;
  },
};
</script>