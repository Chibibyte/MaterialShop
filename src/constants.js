export default {
    BUTTONICONS: {
        AUTHENTICATION: 'mdi-login-variant',
        ADDRESS: 'mdi-package-variant',
        BANKACCOUNT: 'mdi-bank',
        SHIPPING: 'mdi-truck',
        PAYMENT: 'mdi-cash',
        SUMMARY: 'mdi-text-box-check-outline',
        CHECKOUT_COMPLETE: 'mdi-check',
        ADD: 'mdi-plus',
        REMOVE: 'mdi-minus',
        CANCEL: 'mdi-window-close',
        CONFIRM: 'mdi-check',
        SHOPPING: 'mdi-shopping',
        SUCCESS: 'mdi-check',
        FAILURE: 'mdi-window-close',
        WARNING: 'mdi-alert-circle-outline'
    }
}