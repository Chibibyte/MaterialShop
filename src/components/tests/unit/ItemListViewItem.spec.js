import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils';
import ItemListViewItem from '@/components/ItemListViewItem.vue';
import constants from '@/constants.js'

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.prototype.$constants = constants;

let stubs = [
    'v-card',
    'v-row',
    'v-col',
    'v-container',
    'v-icon',
    'v-card-title',
    'v-card-subtitle',
    'v-select',
    'v-btn'
]

let wrapper;
let modules;
let store;
let state;
let defItem = {
    title: 'test title',
    price: '2',
    id: 0
};
beforeEach(() => {
    state = {};
    modules = {
        noncheckout: {
            namespaced: true,
            mutations: {
                addItem: jest.fn(),
                deleteItem: jest.fn()
            }
        }
    }
    store = new Vuex.Store({
        state,
        modules
    })
    wrapper = shallowMount(ItemListViewItem, { stubs, store, localVue });
})

describe('Component', () => {
    test('is instantiated', () => {
        expect(wrapper.isVueInstance).toBeTruthy();
    })

    test('default snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
    })
})

describe('method: addItem', () => {
    test('default', () => {
        wrapper.vm.addItem(defItem)
        expect(modules.noncheckout.mutations.addItem.mock.calls).toHaveLength(1);
        expect(modules.noncheckout.mutations.addItem.mock.calls[0][1]).toEqual(defItem);
    })

    test('no item', () => {
        wrapper.vm.addItem();
        expect(modules.noncheckout.mutations.addItem.mock.calls).toHaveLength(1);
        expect(modules.noncheckout.mutations.addItem.mock.calls[0][1]).toEqual(undefined);
    })
})

describe('method: deleteItem', () => {
    test('default', () => {
        wrapper.vm.deleteItem(3)
        expect(modules.noncheckout.mutations.deleteItem.mock.calls).toHaveLength(1);
        expect(modules.noncheckout.mutations.deleteItem.mock.calls[0][1]).toEqual(3);
    })

    test('no item', () => {
        wrapper.vm.deleteItem();
        expect(modules.noncheckout.mutations.deleteItem.mock.calls).toHaveLength(1);
        expect(modules.noncheckout.mutations.deleteItem.mock.calls[0][1]).toEqual(undefined);
    })
})

describe('v-select', () => {
    test('emit 3', async () => {
        wrapper.vm.amount = 3;
        await wrapper.vm.$nextTick();
        expect(wrapper.element).toMatchSnapshot();
    })
})