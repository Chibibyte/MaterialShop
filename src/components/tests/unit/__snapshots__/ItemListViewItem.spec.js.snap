// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component default snapshot 1`] = `<v-hover />`;

exports[`v-select emit 3 1`] = `<v-hover />`;
